keps/sig-node/24-apparmor/README.md

```diff
@@ -0,0 +1,593 @@
+The notable exceptions are that the AppArmor annotations are immutable on pods, which simplifies the
+migration. AppArmor is also feature gated, via the `AppArmor` gate.
+
+### Goals
+
+- Promote AppArmor to GA
+- Fully document and formally spec the feature support
+- Add equivalent API fields to replace AppArmor annotations and provide a pod
+  level field, which applies to all containers.
+- Deprecate the AppArmor annotations
+
+### Non-Goals
+
+This KEP proposes the absolute minimum to get AppArmor to GA, therefore all
+functional enhancements are out of scope, including:
+
+- Defining any standard "Kubernetes branded" AppArmor profiles
+- Formally specifying the AppArmor profile format in Kubernetes
+- Providing mechanisms for loading profiles from outside of the node
+- Changing the semantics around AppArmor support
+- Windows support
+
+## Proposal
+
+AppArmor is not available on every Linux distribution. Beside this, container runtimes have AppArmor
+as compile-time feature which may be disabled as well. With the GA API the error handling will be
+unchanged, and behave exactly the same as the current error propagation paths (see
+[Failure and Fallback Strategy](#failure-and-fallback-strategy)).
+
+### API
+
+The AppArmor API will be functionally equivalent to the current beta API, with
+the enhancement of adding pod level profiles to match the behavior with seccomp.
+This includes the Pod API, which specifies what profile the containers run with.
+
+#### Pod API
+
+The Pod AppArmor API is generally immutable, except in `PodTemplates`.
+
+```go
+type PodSecurityContext struct {
+    ...
+    // The AppArmor options to use by the containers in this pod.
+    // Note that this field cannot be set when spec.os.name is windows.
+    // +optional
+    AppArmorProfile  *AppArmorProfile
+    ...
+}
+
+type SecurityContext struct {
+    ...
+    // The AppArmor options to use by this container. If AppArmor options are
+    // provided at both the pod & container level, the container options
+    // override the pod options.
+    // Note that this field cannot be set when spec.os.name is windows.
+    // +optional
+    AppArmorProfile  *AppArmorProfile
+    ...
+}
+
+// AppArmorProfile defines a pod or container's AppArmor settings.
+// Only one profile source may be set.
+// +union
+type AppArmorProfile struct {
+    // type indicates which kind of AppArmor profile will be applied.
+    // Valid options are:
+    //   Localhost - a profile pre-loaded on the node.
+    //   RuntimeDefault - the container runtime's default profile.
+    //   Unconfined - no AppArmor enforcement.
+    // +unionDescriminator
+    Type AppArmorProfileType
+
+    // LocalhostProfile indicates a loaded profile on the node that should be used.
+    // The profile must be preconfigured on the node to work.
+    // Must match the loaded name of the profile.
+    // Must only be set if type is "Localhost".
```

Comment:
I wonder if this https://github.com/kubernetes/kubernetes/pull/117020 applies here equally? Sorry I didn't check, maybe you can answer off the top of your head. 

Also wording may be improved to say "must be set when type is localhost and must not be set in any other case"

Reply:
Seccomp already applies validation that LocalhostProfile is set IFF the type is Localhost: https://github.com/kubernetes/kubernetes/blob/0f817dcd6577b663f6cb22cd5080f14138e8b6ac/pkg/apis/core/validation/validation.go#L4612-L4622

I'm not sure the tightening validation was handled correctly for that, but it's been merged since 1.19 at this point.

Reply:
I believe in seccomp case it was an actual security problem (minor), this is why the IFF was added.